File: custom_dir/custom_actions/random_touch.py
# -*- coding: UTF-8 -*-
from maa.context import Context
from maa.custom_action import CustomAction
import random
from maa.agent.agent_server import AgentServer

@AgentServer.custom_action("RandomTouch")
class RandomTouch(CustomAction):
    def run(self,
            context: Context,
            argv: CustomAction.RunArg, ) -> bool:
        print("开始执行自定义动作：随机点击")
        x, y, w, h = argv.box.x, argv.box.y, argv.box.w, argv.box.h
        center_x = x + w / 2
        center_y = y + h / 2
        random_x = random.gauss(center_x, w / 6)
        random_y = random.gauss(center_y, h / 6)
        random_x = min(max(random_x, x), x + w)
        random_y = min(max(random_y, y), y + h)
        context.tasker.controller.post_click(round(random_x), round(random_y)).wait()
        print(f"随机生成的点: {random_x, random_y}")
        return True

    def stop(self) -> None:
        pass

File: agent/actions/random_touch/random_touch.go
package random_touch

import (
	"fmt"
	"math/rand"
	"time"

	"github.com/MaaXYZ/maa-framework-go/v4"
)

type RandomTouch struct{}

// Run 执行随机点击操作
func (a *RandomTouch) Run(ctx *maa.Context, arg *maa.CustomActionArg) bool {
	fmt.Println("开始执行自定义动作：随机点击")

	x := arg.Box.X()
	y := arg.Box.Y()
	w := arg.Box.Width()
	h := arg.Box.Height()

	centerX := float64(x) + float64(w)/2
	centerY := float64(y) + float64(h)/2

	// 使用正态分布生成随机点
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	randomX := r.NormFloat64()*(float64(w)/6) + centerX
	randomY := r.NormFloat64()*(float64(h)/6) + centerY

	// 边界检查
	minVal := func(a, b float64) float64 {
		if a < b {
			return a
		}
		return b
	}
	maxVal := func(a, b float64) float64 {
		if a > b {
			return a
		}
		return b
	}

	randomX = maxVal(float64(x), minVal(randomX, float64(x+w)))
	randomY = maxVal(float64(y), minVal(randomY, float64(y+h)))

	// 执行点击
	controller := ctx.GetTasker().GetController()
	if controller != nil {
		controller.PostClick(int32(randomX), int32(randomY)).Wait()
	}

	fmt.Printf("随机生成的点: (%.2f, %.2f)\n", randomX, randomY)
	return true
}